File: runtime/src/main/java/org/apache/flink/agents/runtime/skill/SkillSourceRegistry.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.agents.runtime.skill;

import org.apache.flink.agents.api.skills.SkillUrlUtils;
import org.apache.flink.agents.runtime.skill.repository.ClasspathSkillRepository;
import org.apache.flink.agents.runtime.skill.repository.FileSystemSkillRepository;
import org.apache.flink.agents.runtime.skill.repository.URLSkillRepository;

import java.util.Locale;
import java.util.Map;
import java.util.TreeSet;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

/**
 * Scheme-keyed registry mapping a skill source scheme (e.g. {@code "local"}, {@code "url"}, {@code
 * "classpath"}) to a {@link SkillSourceHandler}. Built-ins are registered via the static
 * initializer; external code may add custom schemes via {@link #register}.
 *
 * <p>Adding a new source = one {@link #register} call plus one {@code SkillRepository}
 * implementation. {@code SkillManager} and {@code AgentPlan} need not change.
 */
public final class SkillSourceRegistry {

    private static final Map<String, SkillSourceHandler> HANDLERS = new ConcurrentHashMap<>();

    static {
        register(
                "local",
                (params, cl) -> new FileSystemSkillRepository(require(params, "local", "path")),
                params -> params.getOrDefault("path", ""));
        register(
                "url",
                (params, cl) ->
                        new URLSkillRepository(
                                require(params, "url", "url"),
                                params.get("sha256"),
                                Boolean.parseBoolean(
                                        params.getOrDefault("allow_insecure_http", "false"))),
                params -> SkillUrlUtils.redact(params.get("url")));
        register(
                "classpath",
                (params, cl) ->
                        new ClasspathSkillRepository(require(params, "classpath", "resource"), cl),
                params -> params.getOrDefault("resource", ""));
    }

    private SkillSourceRegistry() {}

    /**
     * Register a {@code handler} under {@code scheme}. The scheme is normalized to lowercase. An
     * existing handler for the same scheme is replaced. {@link
     * SkillSourceHandler#describeLocation(Map)} falls back to the default (raw {@code params}); use
     * {@link #register(String, SkillSourceHandler, Function)} to supply a tighter description.
     */
    public static void register(String scheme, SkillSourceHandler handler) {
        HANDLERS.put(scheme.toLowerCase(Locale.ROOT), handler);
    }

    /**
     * Register a {@code handler} under {@code scheme} together with a {@code describer} that
     * extracts the human-readable location from {@code params} (e.g. the {@code path} or {@code
     * url} value). The two functions are composed into one {@link SkillSourceHandler} so that
     * {@link SkillManager} can ask the handler "where did this come from?" without a parallel
     * scheme ladder.
     */
    public static void register(
            String scheme,
            SkillSourceHandler handler,
            Function<Map<String, String>, String> describer) {
        register(
                scheme,
                new SkillSourceHandler() {
                    @Override
                    public SkillRepository open(Map<String, String> params, ClassLoader cl)
                            throws java.io.IOException {
                        return handler.open(params, cl);
                    }

                    @Override
                    public String describeLocation(Map<String, String> params) {
                        return describer.apply(params);
                    }
                });
    }

    /**
     * Return the handler for {@code scheme}, or throw if unknown. Throwing message lists the
     * currently registered schemes to aid debugging cross-language plan mismatches.
     */
    public static SkillSourceHandler get(String scheme) {
        SkillSourceHandler handler = HANDLERS.get(scheme.toLowerCase(Locale.ROOT));
        if (handler == null) {
            throw new IllegalArgumentException(
                    "Unknown skill source scheme: "
                            + scheme
                            + ". Registered schemes: "
                            + new TreeSet<>(HANDLERS.keySet()));
        }
        return handler;
    }

    private static String require(Map<String, String> params, String scheme, String key) {
        String value = params.get(key);
        if (value == null) {
            throw new IllegalArgumentException(
                    "Missing required param '"
                            + key
                            + "' for skill source scheme '"
                            + scheme
                            + "'");
        }
        return value;
    }
}

File: python/flink_agents/runtime/skill/skill_source_registry.py
################################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
#################################################################################
"""Scheme-keyed registry mapping a skill source scheme (e.g. ``"local"``,
``"url"``, ``"package"``) to a handler that opens a :class:`SkillRepository`.
Built-ins are registered at import time; external code may add custom schemes
via :func:`register`.

Adding a new source = one :func:`register` call plus one ``SkillRepository``
implementation. ``SkillManager`` need not change.

The Java side has an analogous ``SkillSourceRegistry`` registering ``local``
/ ``url`` / ``classpath``. Cross-language unsupported schemes (Java's
``classpath`` on Python, Python's ``package`` on Java) hit the registry's
fail-loud path at load time.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Callable, Dict, Mapping

from flink_agents.api.skills import redact_skill_url
from flink_agents.runtime.skill.repository.filesystem_repository import (
    FileSystemSkillRepository,
)
from flink_agents.runtime.skill.repository.package_repository import (
    PackageSkillRepository,
)
from flink_agents.runtime.skill.repository.url_repository import URLSkillRepository

if TYPE_CHECKING:
    from flink_agents.runtime.skill.skill_repository import SkillRepository


@dataclass(frozen=True)
class SkillSourceHandler:
    """Pair of (open, describe_location) bound to a scheme.

    ``describe_location`` returns the human-readable source location for
    :class:`SkillOrigin`. The default falls back to the raw params dict;
    built-ins override it to point at the relevant param (e.g. ``path`` for
    local). Keeping description on the handler removes the parallel scheme
    ladder ``SkillManager`` would otherwise need.
    """

    open: Callable[[Mapping[str, str]], SkillRepository]
    describe_location: Callable[[Mapping[str, str]], str] = field(
        default=lambda params: str(dict(params)),
    )


_HANDLERS: Dict[str, SkillSourceHandler] = {}


def register(
    scheme: str,
    open_fn: Callable[[Mapping[str, str]], SkillRepository],
    describe_location: Callable[[Mapping[str, str]], str] | None = None,
) -> None:
    """Register a handler under ``scheme``. Scheme is normalized to lowercase.

    ``describe_location`` defaults to the raw params dict; pass a tighter
    function (e.g. ``lambda p: p.get("path", "")``) to produce a clean origin.
    """
    handler = (
        SkillSourceHandler(open=open_fn)
        if describe_location is None
        else SkillSourceHandler(open=open_fn, describe_location=describe_location)
    )
    _HANDLERS[scheme.lower()] = handler


def get(scheme: str) -> SkillSourceHandler:
    """Return the handler for ``scheme``, or raise if unknown.

    The error message lists the currently registered schemes to aid debugging
    cross-language plan mismatches.
    """
    handler = _HANDLERS.get(scheme.lower())
    if handler is None:
        msg = (
            f"Unknown skill source scheme: {scheme}. "
            f"Registered schemes: {sorted(_HANDLERS.keys())}"
        )
        raise ValueError(msg)
    return handler


def _require(params: Mapping[str, str], scheme: str, key: str) -> str:
    value = params.get(key)
    if value is None:
        msg = f"Missing required param '{key}' for skill source scheme '{scheme}'"
        raise ValueError(msg)
    return value


register(
    "local",
    lambda params: FileSystemSkillRepository(_require(params, "local", "path")),
    lambda params: params.get("path", ""),
)
register(
    "url",
    lambda params: URLSkillRepository(
        _require(params, "url", "url"),
        sha256=params.get("sha256"),
        allow_insecure_http=params.get("allow_insecure_http", "false").lower()
        == "true",
    ),
    lambda params: redact_skill_url(params.get("url", "")),
)
register(
    "package",
    lambda params: PackageSkillRepository(
        _require(params, "package", "package"),
        _require(params, "package", "resource"),
    ),
    lambda params: f"{params.get('package', '')}/{params.get('resource', '')}",
)
